runtime/common/metering.go

```diff
@@ -217,121 +232,255 @@ func max(a, b int) int {
-		),
+		resultWordLength * bigIntWordSize,
 	)
 }
 
 func NewDivBigIntMemoryUsage(a, b *big.Int) MemoryUsage {
-	return NewBigIntMemoryUsage(
-		// TODO: https://github.com/dapperlabs/cadence-private-issues/issues/32
-		max(
-			BigIntByteLength(a),
-			BigIntByteLength(b),
-		),
-	)
+	return NewModBigIntMemoryUsage(a, b)
 }
 
 func NewBitwiseOrBigIntMemoryUsage(a, b *big.Int) MemoryUsage {
+	// if a >= 0 and b >= 0:
+	//     max(|a|, |b|) + 4
+	// else if a <= 0 and b <= 0:
+	//     |a| + |b| + min(|a|, |b|) + 13
+	// else:
+	//     2 * max(|a|, |b|) + 9
+
+	aWordLength := len(a.Bits())
+	bWordLength := len(b.Bits())
+
+	var resultWordLength int
+	if a.Sign() >= 0 && b.Sign() >= 0 {
+		resultWordLength = max(aWordLength, bWordLength) + 4
+	} else if a.Sign() <= 0 && b.Sign() <= 0 {
+		resultWordLength = aWordLength + bWordLength + min(aWordLength, bWordLength) + 13
+	} else {
+		resultWordLength = 2*max(aWordLength, bWordLength) + 9
+	}
 	return NewBigIntMemoryUsage(
-		// TODO: https://github.com/dapperlabs/cadence-private-issues/issues/32
-		max(
-			BigIntByteLength(a),
-			BigIntByteLength(b),
-		),
+		resultWordLength * bigIntWordSize,
 	)
 }
 
 func NewBitwiseXorBigIntMemoryUsage(a, b *big.Int) MemoryUsage {
+	// if a >= 0 and b >= 0:
+	//     max(|a|, |b|) + 4
+	// else if a <= 0 and b <= 0:
+	//     |a| + |b| + min(|a|, |b|) + 12
+	// else:
+	//     2 * max(|a|, |b|) + 9
+
+	aWordLength := len(a.Bits())
+	bWordLength := len(b.Bits())
+
+	var resultWordLength int
+	if a.Sign() >= 0 && b.Sign() >= 0 {
+		resultWordLength = max(aWordLength, bWordLength) + 4
+	} else if a.Sign() <= 0 && b.Sign() <= 0 {
+		resultWordLength = aWordLength + bWordLength + min(aWordLength, bWordLength) + 12
+	} else {
+		resultWordLength = 2*max(aWordLength, bWordLength) + 9
+	}
 	return NewBigIntMemoryUsage(
-		// TODO: https://github.com/dapperlabs/cadence-private-issues/issues/32
-		max(
-			BigIntByteLength(a),
-			BigIntByteLength(b),
-		),
+		resultWordLength * bigIntWordSize,
 	)
 }
 
 func NewBitwiseAndBigIntMemoryUsage(a, b *big.Int) MemoryUsage {
+	// if a >= 0 and b >= 0:
+	//     max(|a|, |b|) + 4
+	// else if a <= 0 and b <= 0:
+	//     |a| + |b| + max(|a|, |b|) + 13
+	// else:
+	//     2 * max(|a|, |b|) + 8
+
+	aWordLength := len(a.Bits())
+	bWordLength := len(b.Bits())
+
+	var resultWordLength int
+	if a.Sign() >= 0 && b.Sign() >= 0 {
+		resultWordLength = max(aWordLength, bWordLength) + 4
+	} else if a.Sign() <= 0 && b.Sign() <= 0 {
+		resultWordLength = aWordLength + bWordLength + max(aWordLength, bWordLength) + 13
+	} else {
+		resultWordLength = 2*max(aWordLength, bWordLength) + 8
+	}
 	return NewBigIntMemoryUsage(
-		// TODO: https://github.com/dapperlabs/cadence-private-issues/issues/32
-		max(
-			BigIntByteLength(a),
-			BigIntByteLength(b),
-		),
+		resultWordLength * bigIntWordSize,
 	)
 }
 
+var invalidLeftShift = errors.New("invalid left shift of non-Int64")
+
 func NewBitwiseLeftShiftBigIntMemoryUsage(a, b *big.Int) MemoryUsage {
```

Comment:
I don't think the `shift` value should be a big integer, a `uint` is more than enough.
Then in the metering `b/word_size` would avoid using a big integer division ( I guess the metering itself should be very light and should not allocate memory)